### PR TITLE
Decouple persisting logic from saveEntityRecord

### PR DESCRIPTION
**Pull request by adamziel**
This PR aims to move the post-specific logic to the editor package.

The approach it takes is decoupling the data-layer logic (locking, applying edits etc) from the API logic from (sending a request, acknowledging the result) in `saveEntityRecord`. Instead of having a boolean `autosave` argument, I propose a callback argument called `persist`. Now it's up to the caller to choose the appropriate flavor of persisting. Regular save? Sure! Autosave? Absolutely! Writing to a web socket instead of an API? You got it!

As usually, there are upsides and downsides

The good:
* No more post-specific logic in `core-data` (alright, there's still `__unstablePrePersist`, but it should be easy to get rid of too)
* `saveEntityRecord` is more generic so we shouldn't have more domain-specific logic creeping in
* We're decoupled from the API 

The bad:
* `persist` callback is responsible for its own `receiveEntityRecords` which seems off
* Saving logic is now split over multiple files
* Do we need this much decoupling? Maybe retaining `isAutosave` with two more options (`preprocessAutosaveRequestData`, `preprocessAutosaveResponseData`) would do the trick just fine?
* Current E2E failures are about knowing whether the save currently in progress is an autosave or not which is something `core-data` should expose provided autosaving is a `core-data` feature.

### CODE REVIEW

**Comment by adamziel** [comment]
The tests are failing, but I think we're in a good place to discuss the basic idea behind this PR. cc @noisysocks @youknowriad @gziolo @jsnajdr

**Review by youknowriad — COMMENTED** [review]
I'm still wrapping my head around this and what it means for `core-data` but I think there's something that bothers me, while I like the extraction of post specific logic from `saveEntityRecord`. I don't like the move of saving logic to `editor` package.

In my mental modal `core-data` is the layer of abstraction for the server and when I make edits and I want to trigger an autosave for post, I should do just that `autosavePost()` but not tell it how to autosave the post.

File: packages/core-data/src/actions.js [hunk 1]
```diff
@@ -364,158 +362,31 @@ export const saveEntityRecord = (
...
 			);
-
-			if ( isAutosave ) {
-				// Most of this autosave logic is very specific to posts.
-				// This is fine for now as it is the only supported autosave,
-				// but ideally this should all be handled in the back end,
-				// so the client just sends and receives objects.
-				const currentUser = select.getCurrentUser();
-				const currentUserId = currentUser ? currentUser.id : undefined;
-				const autosavePost = select.getAutosave(
```
Comment (jsnajdr) [inline packages/core-data/src/actions.js:413]: I think that #33201 introduced a bug here. This code wants to fetch the autosave for the given post and wait until it's fetched. Therefore the call should be:
```js
const autosavePost = await resolveSelect.getAutosave( ... );
```

File: packages/core-data/src/actions.js [hunk 2]
```diff
@@ -534,6 +404,32 @@ export const saveEntityRecord = (
...
+
+				await dispatch.editEntityRecord(
+					kind,
+					name,
+					recordId,
+					{
+						[ key ]: evaluatedValue,
+					},
+					{ undoIgnore: true }
+				);
```
Comment (jsnajdr) [inline packages/core-data/src/actions.js:425]: There's one thing I don't understand about this evaluation code: when I perform `saveEntityRecord` where there are both direct and lazy edits:
```js
saveEntityRecord( kind, name, {
  title: ( oldTitle ) => oldTitle + ' (new)',
  content: 'Hello',
}
```
Then `editEntityRecord` will be called with the new `title`, but not with `content`. Why is the `edits` state updated in such seemingly arbitrary way?

File: packages/core-data/src/reducer.js [hunk 3]
```diff
@@ -281,7 +281,6 @@ function entity( entityConfig ) {
 								pending:
 									action.type === 'SAVE_ENTITY_RECORD_START',
 								error: action.error,
-								isAutosave: action.isAutosave,
```
Comment (jsnajdr) [inline packages/core-data/src/reducer.js:284]: This removal will break the `isAutosavingEntityRecord` selector. Which is not currently used by any Gutenberg code, but is a part of public API.

**Review by jsnajdr — COMMENTED** [review]
I like the idea of extracting some helper functions from the `saveEntityRecord` function. That function is now 200 lines long and difficult to navigate.

But I think that exposing functions like `persistAutosaveToAPI` as public API is not sufficiently justified. They are abstracting away a part of the saving logic, but this abstraction is used at only one place. If it doesn't have multiple usages, how do we know it's the right abstraction? I would rather create new APIs only when there is clear demand for them.

File: packages/core-data/src/actions.js (hunk 1 above)
Comment (adamziel) [reply packages/core-data/src/actions.js:413]: Great catch! I started https://github.com/WordPress/gutenberg/pull/34584 to fix this

**Comment by adamziel** [comment]
This PR is now outdated and I don't have any better ideas at the moment. Let's close it.

**Pull request closed without merging**